service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java

```diff
@@ -516,7 +516,7 @@ public void registerLivenessChecks(final HealthCheckHandler handler) {
 
      * @return Future The future object carrying the payload of the credentials get operation, if successful.
      */
-    protected Future<String> validateCredentialsForDevice(final String tenantId, final String type, final String authId,
+    public Future<String> validateCredentialsForDevice(final String tenantId, final String type, final String authId,
```

Comment:
See comment above, please not make this method public.

Reply:
Ok so I have 2 options here:

1. make it public
2. make it protected and then I have to change the unit tests significantly and make then pretty complicated, unless there is a way to mock the credentials client nicely that I don't know about.

Here is a snippet of how I was able to mock the client validation response, it is not pretty:

```
        doAnswer(invocation -> {
            CredentialsResult credResult = CredentialsResult.from(200, createValidCredentialsObject());
            invocation.getArgumentAt(2, Handler.class).handle(Future.succeededFuture(credResult));
            return null;

        }).when(credentialsClient).get(anyString(), anyString(), any(Handler.class));

        doAnswer(invocation -> {
            invocation.getArgumentAt(1, Handler.class).handle(Future.succeededFuture(credentialsClient));
            return null;
        }).when(honoClient).getOrCreateCredentialsClient(anyString(), any(Handler.class));
```

With that I have the problem that CredentialsUtils.findAppropriateValidators does not have access to any secretValidators in my tests so I need to figure a way out to insert one, but that is a lot of boilerplate for some simple tests...


Is there an approach that I missed to mock the validateCredentialsForDevice method or a nicer way to test this?

Reply:
Hi @alexmabsi , I completely see your point now. The snippet is showing code where I agree that this is getting overly complex.

Since I do not want to have this method public, I pushed a compromise solution that is forcing the protocol adapters to implement this method and use the previous method from the base class (see in master branch). 
This is IMHO a solution that makes your approach (spy on the adapter class and instruct mockito how to behave for this method) feasible without making the method public.

Please consider rebasing the PR so your (without a doubt valuable) test works with the protected version of the method.

Reply:
Great thanks! The solution works for me, I'll implement the changes right away